### ISSUE TITLE
Migrate existing Whitehall attachments to Asset Manager

### ISSUE THREAD

**Issue by chrisroos** [alphagov/asset-manager#441]
We can do this once we're saving all new attachments to Asset Manager.

## Prerequisites

- [x] Save new Whitehall attachments to Asset Manager #440

## Monitoring

* [Whitehall queue](https://grafana.publishing.service.gov.uk/dashboard/db/sidekiq-older-versions-of-govuk_sidekiq?refresh=1m&orgId=1&var-Application=whitehall&var-Queues=All)
* [Asset Manager queue](https://grafana.publishing.service.gov.uk/dashboard/file/sidekiq.json?refresh=1m&orgId=1&var-Application=asset-manager&var-Queues=All)

## Tasks

- [x] Determine number of attachment_data files we're expecting to upload to asset-manager
- [x] Enable this task to be batched up
- [x] Run `asset_manager:migrate_attachments[0,100000]`
- [x] Run `asset_manager:migrate_attachments[100000,200000]`
- [x] Run `asset_manager:migrate_attachments[200000,300000]`
- [x] Run `asset_manager:migrate_attachments[300000,400000]`
- [x] Run `asset_manager:migrate_attachments[400000,600000]`
- [x] Run `asset_manager:migrate_attachments[600000,700000]`
- [x] Run `asset_manager:migrate_assets[system/uploads/attachment_data/file/675108]`
- [x] Run `asset_manager:migrate_assets[system/uploads/attachment_data/file/681190]`
- [x] Run `asset_manager:migrate_assets[system/uploads/attachment_data/file/681595]`
- [x] Run `asset_manager:migrate_assets[system/uploads/attachment_data/file/681619]`
- [ ] Check that all of the files have been added to Asset Manager.

**Comment by chrisroos** [comment]
I'm going to start working on this.

**Comment by chrisroos** [comment]
In order to be able to migrate these attachments before adding the access-limiting functionality to Asset Manager (#450) we'll need to make sure that it's not possible to access these assets via Asset Manager. I can think of two ways to do that:

1. Mark all the existing attachments as draft when uploading them.
    * This matches the behaviour of new Whitehall attachments being uploaded to Asset Manager (https://github.com/alphagov/whitehall/pull/3734).
    * This would restrict access to these attachments to authenticated users (assuming it were even possible to request them from asset-manager).
    * We'll have to mark the vast majority of these attachments as public in Asset Manager before we can start serving them from Asset Manager.
2. Rely on nginx configuration to prevent access to these assets.
    * This would rely on us not serving these Whitehall attachments from Asset Manager until we've got all the access-limiting functionality in place.
    * We'll have to mark a small number of these attachments as draft in Asset Manager before we can start serving them from Asset Manager.

@floehopper / @chrislo: Do you have any thoughts about either of these approaches, and/or can you think of another way we can start work on this?

**Comment by chrislo** [comment]
I think keeping the migration as close as possible to how new assets are added sounds good. Option 1 sounds like it would work. We can hook the publishing events up before we start serving them too and make sure that is working as expected. Do you think marking a large number of assets as public will be difficult?

**Comment by floehopper** [comment]
I concur. Let's go with option 1. I don't think marking a load of the assets as public later should be too difficult.

**Comment by chrisroos** [comment]
Thanks @chrislo and @floehopper. I'll go with option 1.

**Comment by chrisroos** [comment]
The changes in https://github.com/alphagov/whitehall/pull/3742 were merged and deployed yesterday afternoon and I've just confirmed that we're uploading new attachments in Whitehall to Asset Manager (see #440).

**Comment by chrisroos** [comment]
I've asked 2ndline to run `find /mnt/uploads/whitehall/clean/system/uploads/attachment_data/ -type f | wc -l` in production so that we have an idea of the number of jobs we're expecting to put on the queue in Whitehall.

**Comment by chrisroos** [comment]
I've been thinking about what we might do if the number of jobs starts causing us a problem.

Although the jobs will be added to the `asset_migration` queue, that queue is processed by the same worker as all the other queues in Whitehall (albeit at a lower priority). In the event that processing the `asset_migration` queue causes problems I _think_ we'd want to stop the Whitehall sidekiq workers, remove the `asset_migration` queue from Whitehall's sidekiq.yaml config, redeploy Whitehall and restart the workers. That will give us time to work out how to actually solve the problem.

**Comment by chrisroos** [comment]
2ndline ran the `find` command in production to count the number of `attachment_data` files we're expecting to migrate:

```
find: '/mnt/uploads/whitehall/clean/system/uploads/attachment_data/file/462737': Permission denied
1425876
```

**Comment by chrisroos** [comment]
We asked 2ndline to run the `asset_manager:migrate_attachments` Rake task in production. It pushed about 1,000,000 jobs onto the queue before Redis ran out of RAM. We've manually deleted 800,000 of these jobs and will need to add some kind of batching before we can run this job again.

**Comment by chrisroos** [comment]
I opened https://github.com/alphagov/whitehall/pull/3753 to ensure the asset_migration queue has the lowest priority. This was deployed yesterday afternoon. The graph below shows that it's working as expected because the `asset_migration` jobs were paused while the `link_checks` were running between about 2am and 6am.

![screen shot 2018-02-03 at 08 24 56](https://user-images.githubusercontent.com/6556/35765448-c2d8b736-08bb-11e8-8f4b-8a88d326b676.png)

**Comment by chrislo** [comment]
The `asset_migration` queue was empty at around 14:00 on 2018-02-03. 

<img width="580" alt="screen shot 2018-02-05 at 07 43 45" src="https://user-images.githubusercontent.com/16707/35805084-7643d3e0-0a48-11e8-9d37-fcf974de3c95.png">

**Comment by chrislo** [comment]
The changes to the task to allow us to batch up the migration has been deployed.

**Comment by chrislo** [comment]
I've asked 2ndline to run the rake task to migrate the first batch. I may change the batch sizes after this first one has run and we can see how long it takes.

**Comment by chrislo** [comment]
The number of jobs (i.e. files) on the queue peaked at around 220k with the first batch size of 100k directories.

**Comment by chrislo** [comment]
With the pause over-night (while link checker runs and given there's only a single worker) it looks like it takes around 24hrs to process a batch of 100,000 directories. 

<img width="1289" alt="screen shot 2018-02-06 at 09 07 20" src="https://user-images.githubusercontent.com/16707/35863643-4bcaf406-0b1d-11e8-8d61-9aee8d69f4e7.png">

I'm going to ask 2ndline to run the task to queue the second batch.

**Comment by chrislo** [comment]
I asked second line to run `asset_manager:migrate_attachments[400000,600000]` to queue up a double batch over the weekend. As of now (14:50) there are 508k jobs on the queue.

**Comment by chrislo** [comment]
I checked in integration for the largest directory number (`ls -1 /data/uploads/whitehall/clean/system/uploads/attachment_data/file/ | sort -n | tail`). It is `680559`. I've asked 2ndline run the final batch. There are currently ~215k jobs on the queue.

**Comment by chrislo** [comment]
The queue is now empty and all of the rake task batches have been run.

**Comment by chrislo** [comment]
The migration took around 8 days in total: 

<img width="1335" alt="screen shot 2018-02-13 at 11 26 22" src="https://user-images.githubusercontent.com/16707/36161159-fac502ec-10b0-11e8-9f7a-6ee018ef94df.png">

**Comment by chrislo** [comment]
Grafana reports that there are currently 24 jobs in the Whitehall sidekiq retry set on production compared to 40 jobs at the start of the process 9 days ago. It's not possible to tell whether there are any migration jobs in the retry queue from looking at Grafana, we may want to ask someone to inspect the queue for us in production to make sure.

**Comment by chrislo** [comment]
I think we'll need to run something like 

```
WhitehallAsset.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count

WhitehallAsset.deleted.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count
```

against Asset Manager. And

```
find /data/uploads/whitehall/clean/system/uploads/attachment_data/file -type f | wc -l
```

On the same machine to count the assets on the filesystem.

**Comment by chrislo** [comment]
Second-line ran the commands 

```
WhitehallAsset.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count : 1430728
WhitehallAsset.deleted.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count
: 1018
```

Followed by the find command

```
find /data/uploads/whitehall/clean/system/uploads/attachment_data/file -type f | wc -l
1430333
```

And then the Asset Manager queries again

```
WhitehallAsset.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count
1430756
WhitehallAsset.deleted.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count
1018
```

**Comment by chrislo** [comment]
I just noticed that there are some assets with paths that include, but do not start with, `/government/uploads/system/uploads/attachment_data/file/`. This might explain the difference. On integration: 

```
irb(main):001:0> WhitehallAsset.where(legacy_url_path: %r(/government/uploads/system/uploads/attachment_data/file/)).count
=> 1430330
irb(main):002:0> WhitehallAsset.where(legacy_url_path: %r(^/government/uploads/system/uploads/attachment_data/file/)).count
=> 1429919
irb(main):003:0> 1430330 - 1429919
=> 411
```

**Comment by chrislo** [comment]
I noticed that the sync of assets to staging [completed](https://deploy.publishing.service.gov.uk/job/Copy_Data_to_Staging/983/) shortly after midnight this morning. Given that all the assets we migrated were migrated yesterday, running the comparison on staging should be sufficient - and reduces the likelihood that new assets are added while the commands are run. 

We ran the following commands on staging:

```
irb(main):001:0> WhitehallAsset.where(legacy_url_path: %r(^/government/uploads/system/uploads/attachment_data/file/)).count
=> 1429919
```

```
murraysteele@staging-asset-master-1:~$ find /mnt/uploads/whitehall/clean/system/uploads/attachment_data/file -type f | wc -l
find: `/mnt/uploads/whitehall/clean/system/uploads/attachment_data/file/462737': Permission denied
1429923
```

Followed by 

```
irb(main):001:0> WhitehallAsset.where(legacy_url_path: %r(^/government/uploads/system/uploads/attachment_data/file/)).count
=> 1429919
```

So no new assets were added to asset manager while the find command ran. 

The directory that could not be read contains two assets (a PDF and a thumbnail) and has been given the directory permission `drwxrwx---` instead of `drwxrwxr-x` which all the other directories have. I'm not giving the filenames here in case this directory has been made unreadable for a reason. 

The 4 additional assets on the file system could be explained by a difference in the syncing time to staging of the AM DB and the file-system and/or the fact that the 2 assets in this directory were not migrated. I'm still investigating.

**Comment by chrislo** [comment]
To check if the difference of 4 assets was to do with the difference in sync times of the DB and the filesystem we ran the following on staging

```
find /mnt/uploads/whitehall/clean/system/uploads/attachment_data/file -type f -not -newermt "2018-02-15 00:00:00" | wc -l
1430392
```

```
WhitehallAsset.where(legacy_url_path: %r(^/government/uploads/system/uploads/attachment_data/file/)).where(:created_at.lt => Time.parse("2018-02-15 00:00:00")).count
1430388
```

which still shows that there are 4 more assets on the filesystem than in the asset manager DB.

**Comment by chrislo** [comment]
Having ruled out that the missing attachments were due to timing differences in the syncronisation on staging, 2ndline helped me to produce the full list of files on the file system and records of files from the asset manager database. 

We ran 

```
sudo find /mnt/uploads/whitehall/clean/system/uploads/attachment_data/file -type f -not -newermt "2018-02-15 00:00:00" | gzip > /tmp/attachment_data_file_list.gz
```

on the `asset-master` machine in staging. Note the use of `sudo` to include the 2 assets that had unusual file permissions (see above). 

We also ran 

```
q = WhitehallAsset.where(legacy_url_path: %r(^/government/uploads/system/uploads/attachment_data/file/)).where(:created_at.lt => Time.parse("2018-02-15 00:00:00"))

  File.open("/tmp/asset_manager_assets.txt", 'w') do |file|
    q.pluck(:legacy_url_path).each do |path|
      file.puts path
    end
  end; 0
```

in an `asset-manager` rails console in staging.

I unzipped and sorted the two files, and removed the directory prefixes

```
sed 's/\/government\/uploads\/system\/uploads\///' asset_manager_assets.txt | sort > asset_manager_assets_sorted.txt
sed 's/\/mnt\/uploads\/whitehall\/clean\/system\/uploads\///' attachment_data_file_list.txt | sort > attachment_data_file_list_sorted.txt
```

As expected the list of files on disk had 6 more lines (the 4 missing assets identified yesterday plus the additional 2 that were excluded due to incorrect file permissions) than the list of attachments in the AM database

```
$ wc -l asset_manager_assets_sorted.txt
 1430388 asset_manager_assets_sorted.txt
$ wc -l attachment_data_file_list_sorted.txt
 1430394 attachment_data_file_list_sorted.txt
```

I ran `comm` to find the differences

```
$ comm -3 attachment_data_file_list_sorted.txt asset_manager_assets_sorted.txt
attachment_data/file/675108/0017_MID_Issue_6.pdf
attachment_data/file/675108/thumbnail_0017_MID_Issue_6.pdf.png
	attachment_data/file/679997/Concrete_cladding_-_monitoring_building_condition.odt
	attachment_data/file/680055/Concrete_cladding_-_monitoring_building_condition_.odt
attachment_data/file/681190/Pre-release_access_list.pdf
attachment_data/file/681190/thumbnail_Pre-release_access_list.pdf.png
attachment_data/file/681595/HMT_gender_diversity_targets_Feb_2018.pdf
attachment_data/file/681595/thumbnail_HMT_gender_diversity_targets_Feb_2018.pdf.png
attachment_data/file/681619/Surname_Registration_Docs_for_website.pdf
attachment_data/file/681619/thumbnail_Surname_Registration_Docs_for_website.pdf.png
```

As it turns out there are 8 assets on disk that are not in AM, and 2 in AM that are not on the whitehall NFS mount. The directory with the unusual file permissions `attachment_data/file/462737` is not in the list of differences, so it has been migrated successfully (the migration job must have run as a user with sufficient permissions). 

We should be able to migrate the missing assets by running:

```
asset_manager:migrate_assets[system/uploads/attachment_data/file/675108]
asset_manager:migrate_assets[system/uploads/attachment_data/file/681190]
asset_manager:migrate_assets[system/uploads/attachment_data/file/681595]
asset_manager:migrate_assets[system/uploads/attachment_data/file/681619]
```

**Comment by chrislo** [comment]
joskoetsier ran the 4 rake tasks in production just now. I've checked sentry and there hasn't been any exceptions from the jobs in `whitehall` or `asset-manager`. I intend to let the staging sync run over the weekend and check whether we're now in sync on Monday using the procedure above.

**Comment by chrislo** [comment]
I asked joskoetsier to do a check of this again on Staging

```
sudo find /mnt/uploads/whitehall/clean/system/uploads/attachment_data/file -type f -not -newermt "2018-02-19 00:00:00" | wc -l
1431432
```

```
WhitehallAsset.where(legacy_url_path: %r(^/government/uploads/system/uploads/attachment_data/file/)).where(:created_at.lt => Time.parse("2018-02-19 00:00:00")).count
#=> 1431434
```

The difference of 2 extra assets in Asset Manager is expected from [the investigation last week](https://github.com/alphagov/asset-manager/issues/441#issuecomment-365998432). 

I'm now convinced that we've migrated all the attachments across to Asset Manager. 🎉
